File: src/type-system.ts
import type { SpandaType, UnitKind } from "./ast/nodes.js";

export type PhysicalCategory =
  | "scalar"
  | "distance"
  | "duration"
  | "velocity"
  | "acceleration"
  | "angle"
  | "angular_velocity"
  | "mass"
  | "force"
  | "power"
  | "voltage"
  | "current"
  | "temperature"
  | "pressure"
  | "frequency";

const KNOWN_DOMAIN_TYPES = new Set([
  "Mass", "Force", "Power", "Voltage", "Current", "Temperature", "Pressure", "Time",
  "Timestamp", "Interval", "Waypoint", "MotionCommand", "ControlSignal", "PIDConfig", "GpsFix",
  "ImuData", "AudioFrame", "Prompt", "Completion", "Embedding", "Token", "Context", "Memory",
  "Plan", "ReasoningTrace", "Agent", "Goal", "Task", "Skill", "Capability", "Intent", "Command",
  "Conversation", "Speech", "Gesture", "Emotion", "Feedback", "Risk", "Hazard",
  "SafetyConstraint", "Twin", "SimulationState", "Telemetry", "Replay", "Fault", "Scenario",
  "KnowledgeGraph", "Belief", "Observation", "WorldModel", "Policy", "Reward", "StateEstimate",
  "LLM", "VisionModel", "EmbeddingModel", "CameraFrame", "Image", "DepthImage", "PointCloud",
  "LidarScan", "Goal",
]);

function genericArity(name: string): number | undefined {
  switch (name) {
    case "Array":
    case "Set":
    case "Queue":
    case "Stack":
    case "Topic":
    case "Message":
    case "Endpoint":
      return 1;
    case "Map":
    case "Service":
    case "Tuple":
      return 2;
    case "Action":
      return 3;
    default:
      return undefined;
  }
}

export function resolveTypeName(name: string): SpandaType {
  const short = name.replace(/^std\./, "").split(".").pop() ?? name;
  switch (short) {
    case "Int":
    case "int":
      return { kind: "int" };
    case "Float":
    case "float":
      return { kind: "float" };
    case "Bool":
    case "bool":
      return { kind: "bool" };
    case "String":
    case "string":
      return { kind: "string" };
    case "Char":
    case "char":
      return { kind: "char" };
    case "Bytes":
    case "bytes":
      return { kind: "bytes" };
    case "Null":
    case "null":
      return { kind: "null" };
    case "Void":
    case "void":
      return { kind: "void" };
    case "Time":
      return { kind: "named", name: "Time" };
    case "Duration":
      return { kind: "number", unit: "ms" };
    case "Timestamp":
      return { kind: "named", name: "Timestamp" };
    case "Interval":
      return { kind: "named", name: "Interval" };
    case "Distance":
      return { kind: "number", unit: "m" };
    case "Velocity":
      return { kind: "velocity" };
    case "Acceleration":
      return { kind: "number", unit: "m/s²" };
    case "Angle":
      return { kind: "number", unit: "rad" };
    case "AngularVelocity":
      return { kind: "number", unit: "rad/s" };
    case "Mass":
    case "Force":
    case "Power":
    case "Voltage":
    case "Current":
    case "Temperature":
    case "Pressure":
      return { kind: "named", name: short };
    case "Point2D":
    case "Point3D":
    case "Vector2D":
    case "Vector3D":
    case "Quaternion":
    case "Pose":
      return { kind: "pose" };
    case "Transform":
      return { kind: "transform" };
    case "Trajectory":
    case "Path":
      return { kind: "trajectory" };
    case "Waypoint":
    case "MotionCommand":
    case "ControlSignal":
    case "PIDConfig":
      return { kind: "named", name: short };
    case "CameraFrame":
    case "Image":
    case "DepthImage":
    case "PointCloud":
    case "LidarScan":
    case "Scan":
      return { kind: "scan" };
    case "GpsFix":
    case "ImuData":
    case "AudioFrame":
    case "LLM":
    case "VisionModel":
    case "EmbeddingModel":
    case "Prompt":
    case "Completion":
    case "Embedding":
    case "Token":
    case "Context":
    case "Memory":
    case "Plan":
    case "ReasoningTrace":
    case "Agent":
    case "Goal":
    case "Task":
    case "Skill":
    case "Capability":
    case "Intent":
      return { kind: "named", name: short };
    case "ActionProposal":
      return { kind: "named", name: "ActionProposal" };
    case "SafeAction":
      return { kind: "named", name: "SafeAction" };
    case "Command":
    case "Conversation":
    case "Speech":
    case "Gesture":
    case "Emotion":
    case "Feedback":
    case "Risk":
    case "Hazard":
    case "SafetyConstraint":
    case "EmergencyStop":
    case "Twin":
    case "SimulationState":
    case "Telemetry":
    case "Replay":
    case "Fault":
    case "Scenario":
    case "KnowledgeGraph":
    case "Belief":
    case "Observation":
    case "WorldModel":
    case "Policy":
    case "Reward":
    case "StateEstimate":
      return { kind: "named", name: short };
    default:
      if (KNOWN_DOMAIN_TYPES.has(short)) {
        return { kind: "named", name: short };
      }
      throw new Error(`Unknown type '${short}'`);
  }
}

export function resolveGenericType(name: string, args: SpandaType[]): SpandaType {
  const base = name.split(".").pop() ?? name;
  const expected = genericArity(base);
  if (expected === undefined) {
    throw new Error(`Unknown generic type '${base}'`);
  }
  if (args.length !== expected) {
    throw new Error(`Type '${base}' expects ${expected} type argument(s), got ${args.length}`);
  }
  return { kind: "generic", name: base, typeArgs: args };
}

export function physicalCategory(ty: SpandaType): PhysicalCategory {
  switch (ty.kind) {
    case "int":
    case "float":
      return "scalar";
    case "number":
      switch (ty.unit) {
        case "m":
          return "distance";
        case "s":
        case "ms":
          return "duration";
        case "m/s":
          return "velocity";
        case "m/s²":
          return "acceleration";
        case "rad":
        case "deg":
          return "angle";
        case "rad/s":
          return "angular_velocity";
        case "Hz":
          return "frequency";
        default:
          return "scalar";
      }
    case "velocity":
      return "velocity";
    case "pose":
      return "distance";
    case "named":
      switch (ty.name) {
        case "Distance":
          return "distance";
        case "Duration":
        case "Time":
        case "Timestamp":
        case "Interval":
          return "duration";
        case "Velocity":
          return "velocity";
        case "Acceleration":
          return "acceleration";
        case "Angle":
        case "AngularVelocity":
          return "angular_velocity";
        case "Mass":
          return "mass";
        case "Force":
          return "force";
        case "Power":
          return "power";
        case "Voltage":
          return "voltage";
        case "Current":
          return "current";
        case "Temperature":
          return "temperature";
        case "Pressure":
          return "pressure";
        default:
          return "scalar";
      }
    default:
      return "scalar";
  }
}

type BinaryOp = "add" | "sub" | "lt" | "lte" | "gt" | "gte" | "eq" | "neq" | "mul" | "div" | "and" | "or";

const OP_MAP: Record<string, BinaryOp> = {
  "+": "add",
  add: "add",
  "-": "sub",
  sub: "sub",
  "<": "lt",
  lt: "lt",
  "<=": "lte",
  lte: "lte",
  ">": "gt",
  gt: "gt",
  ">=": "gte",
  gte: "gte",
  "==": "eq",
  eq: "eq",
  "!=": "neq",
  neq: "neq",
  "*": "mul",
  mul: "mul",
  "/": "div",
  div: "div",
  and: "and",
  or: "or",
};

export function binaryPhysicalOpAllowed(opLexeme: string, left: SpandaType, right: SpandaType): boolean {
  const op = OP_MAP[opLexeme];
  if (!op) return true;
  const catL = physicalCategory(left);
  const catR = physicalCategory(right);
  switch (op) {
    case "add":
    case "sub":
      if (catL === "scalar" && catR === "scalar") return true;
      return catL === catR && catL !== "scalar";
    case "lt":
    case "lte":
    case "gt":
    case "gte":
    case "eq":
    case "neq":
      return catL === catR;
    case "mul":
    case "div":
      return true;
    case "and":
    case "or":
      return left.kind === "bool" && right.kind === "bool";
    default:
      return true;
  }
}

export function isActionProposalType(ty: SpandaType): boolean {
  return ty.kind === "named" && ty.name === "ActionProposal";
}

export function isSafeActionType(ty: SpandaType): boolean {
  return ty.kind === "named" && ty.name === "SafeAction";
}

export function typeKindName(ty: SpandaType): string {
  switch (ty.kind) {
    case "generic":
      return "generic";
    case "enum_variant":
      return "enum_variant";
    default:
      return ty.kind;
  }
}

File: crates/spanda-core/src/type_system.rs
//! Spanda type system: primitives, generics, physical units, and domain types.

use crate::ast::{SpandaType, UnitKind};
use std::collections::HashMap;

/// Physical dimension for unit algebra (reject e.g. `speed + voltage`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum PhysicalCategory {
    Scalar,
    Distance,
    Duration,
    Velocity,
    Acceleration,
    Angle,
    AngularVelocity,
    Mass,
    Force,
    Power,
    Voltage,
    Current,
    Temperature,
    Pressure,
    Frequency,
}

/// Generic type constructor arity.
#[derive(Debug, Clone, Copy)]
pub struct GenericDef {
    pub name: &'static str,
    pub arity: usize,
    pub namespace: Option<&'static str>,
}

/// Resolve a type name (optionally qualified) to a `SpandaType`.
pub fn resolve_type_name(name: &str) -> Result<SpandaType, String> {
    let name = name.strip_prefix("std.").unwrap_or(name);
    let name = name
        .rsplit('.')
        .next()
        .unwrap_or(name);

    match name {
        // Foundation
        "Int" | "int" => Ok(SpandaType::Int),
        "Float" | "float" => Ok(SpandaType::Float),
        "Bool" | "bool" => Ok(SpandaType::Bool),
        "String" | "string" => Ok(SpandaType::String),
        "Char" | "char" => Ok(SpandaType::Char),
        "Bytes" | "bytes" => Ok(SpandaType::Bytes),
        "Null" | "null" => Ok(SpandaType::Null),
        "Void" | "void" => Ok(SpandaType::Void),
        // Time
        "Time" => Ok(SpandaType::Named {
            name: "Time".into(),
        }),
        "Duration" => Ok(SpandaType::Number {
            unit: UnitKind::Ms,
        }),
        "Timestamp" => Ok(SpandaType::Named {
            name: "Timestamp".into(),
        }),
        "Interval" => Ok(SpandaType::Named {
            name: "Interval".into(),
        }),
        // Physical unit types
        "Distance" => Ok(SpandaType::Number {
            unit: UnitKind::M,
        }),
        "Velocity" => Ok(SpandaType::Velocity),
        "Acceleration" => Ok(SpandaType::Number {
            unit: UnitKind::MPerS2,
        }),
        "Angle" => Ok(SpandaType::Number {
            unit: UnitKind::Rad,
        }),
        "AngularVelocity" => Ok(SpandaType::Number {
            unit: UnitKind::RadPerS,
        }),
        "Mass" | "Force" | "Power" | "Voltage" | "Current" | "Temperature" | "Pressure" => {
            Ok(SpandaType::Named {
                name: name.to_string(),
            })
        }
        // Spatial / robotics
        "Point2D" | "Point3D" | "Vector2D" | "Vector3D" | "Quaternion" | "Pose" => {
            Ok(SpandaType::Pose)
        }
        "Transform" => Ok(SpandaType::Transform),
        "Trajectory" | "Path" => Ok(SpandaType::Trajectory),
        "Waypoint" | "MotionCommand" | "ControlSignal" | "PIDConfig" => Ok(SpandaType::Named {
            name: name.to_string(),
        }),
        // Sensors
        "CameraFrame" | "Image" | "DepthImage" | "PointCloud" | "LidarScan" => Ok(SpandaType::Scan),
        "GpsFix" | "ImuData" | "AudioFrame" => Ok(SpandaType::Named {
            name: name.to_string(),
        }),
        // AI
        "LLM" | "VisionModel" | "EmbeddingModel" | "Prompt" | "Completion" | "Embedding"
        | "Token" | "Context" | "Memory" | "Plan" | "ReasoningTrace" => Ok(SpandaType::Named {
            name: name.to_string(),
        }),
        // Agent / autonomy
        "Agent" | "Goal" | "Task" | "Skill" | "Capability" | "Intent" => Ok(SpandaType::Named {
            name: name.to_string(),
        }),
        "ActionProposal" => Ok(SpandaType::Named {
            name: "ActionProposal".into(),
        }),
        "SafeAction" => Ok(SpandaType::Named {
            name: "SafeAction".into(),
        }),
        // HRI
        "Command" | "Conversation" | "Speech" | "Gesture" | "Emotion" | "Feedback" => {
            Ok(SpandaType::Named {
                name: name.to_string(),
            })
        }
        // Safety
        "Risk" | "Hazard" | "SafetyConstraint" | "EmergencyStop" => Ok(SpandaType::Named {
            name: name.to_string(),
        }),
        // Digital twin
        "Twin" | "SimulationState" | "Telemetry" | "Replay" | "Fault" | "Scenario" => {
            Ok(SpandaType::Named {
                name: name.to_string(),
            })
        }
        // Advanced
        "KnowledgeGraph" | "Belief" | "Observation" | "WorldModel" | "Policy" | "Reward"
        | "StateEstimate" => Ok(SpandaType::Named {
            name: name.to_string(),
        }),
        // Legacy aliases
        "Scan" => Ok(SpandaType::Scan),
        other if is_known_domain_type(other) => Ok(SpandaType::Named {
            name: other.to_string(),
        }),
        other => Err(format!("Unknown type '{other}'")),
    }
}

pub fn resolve_generic_type(name: &str, args: &[SpandaType]) -> Result<SpandaType, String> {
    let base = name.rsplit('.').next().unwrap_or(name);
    let expected = generic_arity(base).ok_or_else(|| format!("Unknown generic type '{base}'"))?;
    if args.len() != expected {
        return Err(format!(
            "Type '{base}' expects {expected} type argument(s), got {}",
            args.len()
        ));
    }
    Ok(SpandaType::Generic {
        name: base.to_string(),
        type_args: args.to_vec(),
    })
}

fn generic_arity(name: &str) -> Option<usize> {
    match name {
        "Array" | "Set" | "Queue" | "Stack" | "Topic" | "Message" => Some(1),
        "Map" | "Service" | "Tuple" => Some(2),
        "Action" => Some(3),
        "Endpoint" => Some(1),
        _ => None,
    }
}

fn is_known_domain_type(name: &str) -> bool {
    KNOWN_DOMAIN_TYPES.contains(&name)
}

const KNOWN_DOMAIN_TYPES: &[&str] = &[
    "Mass", "Force", "Power", "Voltage", "Current", "Temperature", "Pressure", "Time",
    "Timestamp", "Interval", "Waypoint", "MotionCommand", "ControlSignal", "PIDConfig", "GpsFix",
    "ImuData", "AudioFrame", "Prompt", "Completion", "Embedding", "Token", "Context", "Memory",
    "Plan", "ReasoningTrace", "Agent", "Goal", "Task", "Skill", "Capability", "Intent", "Command",
    "Conversation", "Speech", "Gesture", "Emotion", "Feedback", "Risk", "Hazard",
    "SafetyConstraint", "Twin", "SimulationState", "Telemetry", "Replay", "Fault", "Scenario",
    "KnowledgeGraph", "Belief", "Observation", "WorldModel", "Policy", "Reward", "StateEstimate",
    "LLM", "VisionModel", "EmbeddingModel", "CameraFrame", "Image", "DepthImage", "PointCloud",
    "LidarScan",
];

/// Physical category used to reject invalid unit operations.
pub fn physical_category(ty: &SpandaType) -> PhysicalCategory {
    match ty {
        SpandaType::Int | SpandaType::Float => PhysicalCategory::Scalar,
        SpandaType::Number { unit, .. } => match unit {
            UnitKind::M => PhysicalCategory::Distance,
            UnitKind::S | UnitKind::Ms => PhysicalCategory::Duration,
            UnitKind::MPerS => PhysicalCategory::Velocity,
            UnitKind::MPerS2 => PhysicalCategory::Acceleration,
            UnitKind::Rad | UnitKind::Deg => PhysicalCategory::Angle,
            UnitKind::RadPerS => PhysicalCategory::AngularVelocity,
            UnitKind::Hz => PhysicalCategory::Frequency,
            UnitKind::None => PhysicalCategory::Scalar,
        },
        SpandaType::Velocity => PhysicalCategory::Velocity,
        SpandaType::Pose => PhysicalCategory::Distance,
        SpandaType::Named { name } => match name.as_str() {
            "Distance" => PhysicalCategory::Distance,
            "Duration" | "Time" | "Timestamp" | "Interval" => PhysicalCategory::Duration,
            "Velocity" => PhysicalCategory::Velocity,
            "Acceleration" => PhysicalCategory::Acceleration,
            "Angle" | "AngularVelocity" => PhysicalCategory::AngularVelocity,
            "Mass" => PhysicalCategory::Mass,
            "Force" => PhysicalCategory::Force,
            "Power" => PhysicalCategory::Power,
            "Voltage" => PhysicalCategory::Voltage,
            "Current" => PhysicalCategory::Current,
            "Temperature" => PhysicalCategory::Temperature,
            "Pressure" => PhysicalCategory::Pressure,
            _ => PhysicalCategory::Scalar,
        },
        _ => PhysicalCategory::Scalar,
    }
}

/// Returns `None` when the operation is invalid (e.g. distance + duration).
pub fn binary_physical_op_allowed(
    op: crate::ast::BinaryOp,
    left: &SpandaType,
    right: &SpandaType,
) -> bool {
    use crate::ast::BinaryOp;
    let cat_l = physical_category(left);
    let cat_r = physical_category(right);

    match op {
        BinaryOp::Add | BinaryOp::Sub => {
            if cat_l == PhysicalCategory::Scalar && cat_r == PhysicalCategory::Scalar {
                return true;
            }
            cat_l == cat_r && cat_l != PhysicalCategory::Scalar
        }
        BinaryOp::Lt
        | BinaryOp::Lte
        | BinaryOp::Gt
        | BinaryOp::Gte
        | BinaryOp::Eq
        | BinaryOp::Neq => cat_l == cat_r,
        BinaryOp::Mul | BinaryOp::Div => true,
        BinaryOp::And | BinaryOp::Or => {
            matches!(left, SpandaType::Bool) && matches!(right, SpandaType::Bool)
        }
    }
}

pub fn is_action_proposal_type(ty: &SpandaType) -> bool {
    matches!(
        ty,
        SpandaType::Named { name } if name == "ActionProposal"
    )
}

pub fn is_safe_action_type(ty: &SpandaType) -> bool {
    matches!(
        ty,
        SpandaType::Named { name } if name == "SafeAction"
    )
}

pub fn std_namespaces() -> HashMap<&'static str, &'static [&'static str]> {
    let mut m = HashMap::new();
    m.insert(
        "std.time",
        &["Time", "Duration", "Timestamp", "Interval"][..],
    );
    m.insert(
        "std.units",
        &[
            "Distance",
            "Velocity",
            "Acceleration",
            "Angle",
            "AngularVelocity",
            "Mass",
            "Force",
            "Power",
            "Voltage",
            "Current",
            "Temperature",
            "Pressure",
        ][..],
    );
    m.insert(
        "std.spatial",
        &[
            "Point2D",
            "Point3D",
            "Vector2D",
            "Vector3D",
            "Quaternion",
            "Pose",
            "Transform",
            "Trajectory",
            "Path",
            "Waypoint",
        ][..],
    );
    m.insert(
        "std.ai",
        &[
            "LLM",
            "VisionModel",
            "EmbeddingModel",
            "Prompt",
            "Completion",
            "Embedding",
            "Token",
            "Context",
            "Memory",
            "Plan",
            "ReasoningTrace",
        ][..],
    );
    m.insert(
        "std.robotics",
        &[
            "MotionCommand",
            "ControlSignal",
            "PIDConfig",
            "ActionProposal",
            "SafeAction",
            "Agent",
            "Goal",
            "Task",
            "Skill",
            "Capability",
            "Intent",
        ][..],
    );
    m.insert(
        "std.sensors",
        &[
            "CameraFrame",
            "Image",
            "DepthImage",
            "PointCloud",
            "LidarScan",
            "GpsFix",
            "ImuData",
            "AudioFrame",
        ][..],
    );
    m.insert(
        "std.safety",
        &["Risk", "Hazard", "SafetyConstraint", "EmergencyStop", "SafeAction"][..],
    );
    m.insert(
        "std.twin",
        &[
            "Twin",
            "SimulationState",
            "Telemetry",
            "Replay",
            "Fault",
            "Scenario",
        ][..],
    );
    m.insert(
        "std.hri",
        &[
            "Command",
            "Conversation",
            "Speech",
            "Gesture",
            "Emotion",
            "Feedback",
        ][..],
    );
    m
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_unknown_type() {
        assert!(resolve_type_name("NotARealType").is_err());
    }

    #[test]
    fn resolves_generics_with_arity() {
        let goal = SpandaType::Named {
            name: "Goal".into(),
        };
        let err = resolve_generic_type("Array", &[]).unwrap_err();
        assert!(err.contains("expects 1"));
        let ok = resolve_generic_type("Array", &[goal]).unwrap();
        assert!(matches!(ok, SpandaType::Generic { .. }));
    }

    #[test]
    fn rejects_mixed_physical_add() {
        let vel = SpandaType::Velocity;
        let volt = SpandaType::Named {
            name: "Voltage".into(),
        };
        assert!(!binary_physical_op_allowed(
            crate::ast::BinaryOp::Add,
            &vel,
            &volt
        ));
    }
}
